File: triatlon-java/server/src/main/java/triatlon/server/TriatlonService.java
package triatlon.server;

import triatlon.model.Result;
import triatlon.model.activity.AthletePoints;
import triatlon.model.activity.RaceType;
import triatlon.model.person.Athlete;
import triatlon.model.person.Referee;
import triatlon.repository.AthleteRepository;
import triatlon.repository.RaceRepository;
import triatlon.repository.RefereeRepository;
import triatlon.service.TriatlonObserverInterface;
import triatlon.service.TriatlonServiceInterface;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.stream.Collectors;

public class TriatlonService implements TriatlonServiceInterface  {

    private final RaceRepository raceRepository;
    private final RefereeRepository refereeRepository;
    private final AthleteRepository athleteRepository;
    private Map<String, TriatlonObserverInterface> loggedReferees;

    public TriatlonService(RaceRepository raceRepository, RefereeRepository refereeRepository, AthleteRepository athleteRepository) {
        this.raceRepository = raceRepository;
        this.refereeRepository = refereeRepository;
        this.athleteRepository = athleteRepository;
        loggedReferees = new ConcurrentHashMap<>();
    }

    @Override
    public synchronized Referee logInReferee(String email, String password, TriatlonObserverInterface observer) {
        Referee referee = this.refereeRepository.findByEmail(email);
        if (referee == null) return null;
        var isLoggedIn = referee.getEmail().equals(email) && referee.getPassword().equals(password);
        if (isLoggedIn) {
            loggedReferees.putIfAbsent(referee.getEmail(), observer);
        }
        return referee;
    }

    @Override
    public synchronized void logOutReferee(String email) {
        loggedReferees.remove(email);
    }

    @Override
    public synchronized List<Athlete> getAthletes() {
        List<Athlete> athletes = new ArrayList<>();
        this.athleteRepository.findAll().forEach(athletes::add);
        return athletes;
    }

    @Override
    public List<Result> getAthletesWithTotalPoints() {
        List<Result> athletesWithResults = new ArrayList<>();
        this.athleteRepository.findAll().forEach(athlete -> {
            var points = 0;
            for (var value : RaceType.values()) {
                raceRepository.setRaceType(value);
                AthletePoints athletePoints = raceRepository.findByAthleteId(athlete.getId());
                if (athletePoints != null) points += athletePoints.getPoints();
            }
            Result athleteResult = new Result(athlete.getId(), athlete.getFullName(), athlete.getFullNameReversed(), points);
            athletesWithResults.add(athleteResult);
        });
        return athletesWithResults.stream().sorted(Comparator.comparing(Result::getNameReversed)).collect(Collectors.toList());
    }

    @Override
    public void addResult(Referee referee, Integer athleteId, Integer points) {
        this.raceRepository.setRaceType(referee.getRaceType());
        AthletePoints athletePoints = this.raceRepository.findByAthleteId(athleteId);
        if (athletePoints == null) {
            athletePoints = new AthletePoints(athleteId, points);
            this.raceRepository.save(athletePoints);
        } else {
            athletePoints.setPoints(points);
            this.raceRepository.update(athletePoints);
        }
        ExecutorService executor = Executors.newFixedThreadPool(5);
        loggedReferees.forEach((email, observer) -> {
            executor.execute(() -> {
                try {
                    Thread.sleep(1000);
                    observer.resultAdded(getAthletesWithTotalPoints());
                } catch (Exception ex) {
                    ex.printStackTrace();
                }
            });
        });
        executor.shutdown();
    }

    @Override
    public List<Result> getParticipantsWithResultInRace(RaceType raceType) {
        this.raceRepository.setRaceType(raceType);
        List<Result> results = new ArrayList<>();
        this.raceRepository.findAllWithPoints().forEach(athletePoints -> {
            Athlete athlete = this.athleteRepository.findOne(athletePoints.getAthleteId());
            Result result = new Result(athlete.getId(), athlete.getFullName(), athlete.getFullNameReversed(), athletePoints.getPoints());
            results.add(result);
        });
        return results.stream().sorted(Comparator.comparing(Result::getPoints).reversed()).collect(Collectors.toList());
    }

}

File: triatlon-cs/Server/TriatlonService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Repository;
using Repository.DB;
using Service;
using Triatlon.Proto;
using Athlete = Model.Person.Athlete;
using AthletePoints = Model.Activity.AthletePoints;
using RaceType = Model.Activity.RaceType;
using Referee = Model.Person.Referee;
using Result = Model.Result;

namespace Server
{
    public class TriatlonService : ITriatlonService
    {
        private readonly IAthleteRepository _athleteRepository;
        private readonly IRefereeRepository _refereeRepository;
        private readonly RaceRepository _raceRepository;
        private readonly IDictionary<string, ITriatlonObserver> _observers;

        public TriatlonService(IDictionary<string, string> properties)
        {
            _athleteRepository = new AthleteDbRepository(properties);
            _refereeRepository = new RefereeDbRepository(properties);
            _raceRepository = new RaceDbRepository(properties, RaceType.Swimming);
            _observers = new Dictionary<string, ITriatlonObserver>();
        }
        
        public Triatlon.Proto.Referee LogInReferee(string email, string password, ITriatlonObserver observer)
        {
            var referee = _refereeRepository.FindByEmail(email);
            if (referee == null) return null;
            var isLoggedIn = referee.Email.Equals(email) && referee.Password.Equals(password);
            if (isLoggedIn)
            {
                _observers[referee.Email] = observer;
            }
            var refereeProto = new Triatlon.Proto.Referee
            {
                Id = referee.Id,
                FirstName = referee.FirstName,
                LastName = referee.LastName,
                RaceType = ProtoUtils.ConvertRaceType(referee.RaceType),
                Email = referee.Email,
                Password = referee.Password
            };
            return refereeProto;
        }

        public void LogOutReferee(string email)
        {
            _observers.Remove(email);
        }

        public List<Triatlon.Proto.Athlete> GetAthletes()
        {
            var athletes = new List<Triatlon.Proto.Athlete>();
            foreach (var athlete in _athleteRepository.FindAll())
            {
                athletes.Add(ProtoUtils.ConvertAthlete(athlete));
            }

            return athletes;
        }

        public List<Triatlon.Proto.Result> GetAthletesWithTotalPoints()
        {
            var athletesWithResults = new List<Triatlon.Proto.Result>();
            foreach (var athlete in _athleteRepository.FindAll())
            {
                var points = 0;
                foreach (var value in Enum.GetValues(typeof(RaceType)))
                {
                    _raceRepository.RaceType = (RaceType) value;
                    var athletePoints = _raceRepository.FindByAthleteId(athlete.Id);
                    if (athletePoints != null) points += athletePoints.Points;
                }
                var athleteResult = new Result(athlete.Id, athlete.FullName, athlete.FullNameReversed, points);
                athletesWithResults.Add(ProtoUtils.ConvertResult(athleteResult));
            }

            return athletesWithResults.OrderBy(r => r.ReversedName).ToList();
        }

        public void AddResult(Triatlon.Proto.Referee referee, int athleteId, int points)
        {
            _raceRepository.RaceType = ProtoUtils.ConvertBackRaceType(referee.RaceType);
            var athletePoints = _raceRepository.FindByAthleteId(athleteId);
            if (athletePoints == null)
            {
                athletePoints = new AthletePoints(athleteId, points);
                _raceRepository.Save(athletePoints);
            }
            else
            {
                athletePoints.Points = points;
                _raceRepository.Update(athletePoints);
            }

            foreach (var observer in _observers)
            {
                Task.Run(() => _observers[observer.Key].ResultAdded(GetAthletesWithTotalPoints()));
            }
        }

        public List<Triatlon.Proto.Result> GetParticipantsWithResultInRace(Triatlon.Proto.RaceType raceType)
        {
            _raceRepository.RaceType = ProtoUtils.ConvertBackRaceType(raceType);
            var results = (from athletePoints in _raceRepository.FindAllWithPoints() let athlete = _athleteRepository.FindOne(athletePoints.AthleteId) select new Result(athlete.Id, athlete.FullName, athlete.FullNameReversed, athletePoints.Points)).ToList();
            var resultsProto = new List<Triatlon.Proto.Result>();
            foreach (var result in results.OrderByDescending(r => r.Points))
            {
                resultsProto.Add(ProtoUtils.ConvertResult(result));
            }

            return resultsProto;
        }
    }
}